src/Socket/Stream/IPv4.hs

```diff
@@ -1280,6 +1283,62 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
       functionReceiveBetweenMutableByteArraySlice
       [SCK.negativeSliceLength]
 
+interruptibleReceiveBoundedByteStringSlice ::
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> Int -- ^ Maximum number of bytes to receive
+  -> Int -- ^ Offset into the buffer
+  -> IO (Either (ReceiveException 'Interruptible) ByteString) -- ^ Either a socket exception or the number of ByteString we allocated
+interruptibleReceiveBoundedByteStringSlice !abandon !conn !total !off = do
+  marr@(MutableByteArray marr#) <- PM.newPinnedByteArray total
+  interruptibleReceiveBoundedMutableByteArraySlice abandon conn total marr off >>= \case
+    Left err -> pure (Left err)
+    Right len -> pure (Right (BI.PS (FP.ForeignPtr (byteArrayContents# (unsafeCoerce# marr#)) (FP.PlainPtr marr#)) len total))
+
+-- | Receive bytes using the given array. This can be interrupted
+--   by the completion of an 'STM' transaction.
+interruptibleReceiveMutableByteArray ::
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> MutableByteArray RealWorld -- ^ Buffer in which the data is going to be stored
+  -> IO (Either (ReceiveException 'Interruptible) Int) -- ^ Either a socket exception or the number of bytes read
+interruptibleReceiveMutableByteArray !abandon !conn !marr = do
+  !sz <- PM.getSizeofMutableByteArray marr
+  interruptibleReceiveMutableByteArraySlice abandon conn marr 0 sz
+
+-- | Receive bytes using the given array and
+--   starting at the given offset. This can be interrupted by the
+--   completion of an 'STM' transaction.
+interruptibleReceiveMutableByteArraySlice ::
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> MutableByteArray RealWorld -- ^ Buffer in which the data is going to be stored
+  -> Int -- ^ Offset into the buffer
+  -> Int -- ^ Length of slice
+  -> IO (Either (ReceiveException 'Interruptible) Int) -- ^ Either a socket exception or the number of bytes read
```

Comment:
This should return `()`, not `Int`. If it succeeds at all, then `total_bytes === requested_bytes`